### ISSUE TITLE
NDFileHDF5 documentation fails to state that "when" is a valid XML attribute for datasets

### ISSUE THREAD

**Issue by MarkRivers** [areaDetector/ADCore#435]
The NDFileHDF5 documentation states that the XML attribute "when" can be used for "attributes" in the HDF5 file, but not for datasets.
https://areadetector.github.io/master/ADCore/NDFileHDF5.html#xml-defined-file-structure-layout 

The schema file also does not allow "when" for datasets.  

The following XML layout file has when="OnFileWrite" for the NumCaptured dataset.  This works correctly, it will record this PV for each array if when="OnFileWrite" but only for one array if it is "OnFileOpen" or "OnFileClose".  

```
<hdf5_layout>
  <group name="detector">
    <dataset name="data1" source="detector" det_default="true">
        <attribute name="AcquireTime" source="ndattribute" ndattribute="AcquireTime" when="OnFileWrite"></attribute>
   </dataset>
    <dataset name="det_name" source="constant" value="Simulated detector" type="string" when="OnFileWrite">
        <attribute name="AcquireTime" source="ndattribute" ndattribute="AcquireTime" when="OnFileWrite"></attribute>
   </dataset>
   <dataset name="NumCaptured" source="ndattribute" ndattribute="NumCaptured" when="OnFileWrite"/>
  </group>
</hdf5_layout>
```
However, this layout file fails to validate with the supplied xsd file in ADCore.

```
corvette:simDetectorIOC/iocBoot/iocSimDetector>xmllint --noout --schema /home/epics/devel/areaDetector/ADCore/XML_schema/hdf5_xml_layout_schema.xsd hdf5_layout_minimal.xml
hdf5_layout_minimal.xml:6: element dataset: Schemas validity error : Element 'dataset', attribute 'when': The attribute 'when' is not allowed.
hdf5_layout_minimal.xml:9: element dataset: Schemas validity error : Element 'dataset', attribute 'when': The attribute 'when' is not allowed.
hdf5_layout_minimal.xml fails to validate
```

So clearly "when" is allowed for datasets, and both the documentation and the xsd schema file are incorrect.

Is "when" actually allowed for HDF5 attributes in the XML file?  I don't think it is.  Note that in the above layout XML file the dataset data1 has an AcquireTime attribute that has when="OnFileWrite".  I collected a stream of 10 arrays with ADSimDetector.  HDFView shows the following for the data1 dataset.
```
data1 (1836, 38)
    8-bit unsigned integer,    10 x 1024 x 1024
    Number of attributes = 5
        AcquireTime = 0.001
        NDArrayDimBinning = 1,1
        NDArrayDimOffset = 0,0
        NDArrayDimReverse = 0,0
        NDArrayNumDims = 2
```
Note that the attribute AcquireTime is present, but there is a single value, not 10.

So I think the documentation and schema file just have it wrong, "when" is allowed for datasets, but not for attributes?  Can someone confirm this?

**Comment by MarkRivers** [comment]
I have done some more testing and I believe I understand the "when" XML attribute now.

- "when" is allowed for HDF5 attributes.  The allowed values are only "OnFileOpen" and "OnFileClose". The default is "OnFileOpen" if "when" is not specified.  "OnFileWrite" is not allowed because that would cause the attribute have an array of values, and that is not supported.
- "when" is allowed for HDF5 datasets.  The allowed values are "OnFileOpen", "OnFileClose" and "OnFileWrite".  "OnFileWrite" is the default if "when" is not specified.

If others agree with this I will update the documentation and the XSD file to agree with the above.

**Comment by MarkRivers** [comment]
Fixed with fd75436dabefd1d4e14da8294de095581763356d and 0a3f0d99f9f8c0999320054420cf1d2ed3b22300.
